File: selenium-tests/src/main/java/com/timesheet/pages/ReportsPage.java
package com.timesheet.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import java.util.List;

public class ReportsPage extends BasePage {

    private final By pageTitle = By.xpath("//h4[contains(text(), 'Reports')]");
    private final By clientSelect = By.cssSelector(".MuiSelect-select");
    private final By clientMenuItems = By.cssSelector(".MuiMenu-paper .MuiMenuItem-root");
    private final By csvExportButton = By.cssSelector("button[aria-label='Export as CSV'], button svg[data-testid='DescriptionIcon']");
    private final By pdfExportButton = By.cssSelector("button[aria-label='Export as PDF'], button svg[data-testid='PictureAsPdfIcon']");
    private final By totalHoursCard = By.xpath("//p[contains(text(), 'Total Hours')]/..//h4 | //*[contains(text(), 'Total Hours')]/following-sibling::*");
    private final By totalEntriesCard = By.xpath("//p[contains(text(), 'Total Entries')]/..//h4 | //*[contains(text(), 'Total Entries')]/following-sibling::*");
    private final By averageHoursCard = By.xpath("//p[contains(text(), 'Average Hours')]/..//h4 | //*[contains(text(), 'Average Hours')]/following-sibling::*");
    private final By reportTable = By.cssSelector("table");
    private final By reportRows = By.cssSelector("table tbody tr");
    private final By noEntriesMessage = By.xpath("//*[contains(text(), 'No work entries found for this client')]");
    private final By noClientsMessage = By.xpath("//*[contains(text(), 'need to create at least one client')]");
    private final By selectClientMessage = By.xpath("//*[contains(text(), 'Select a client to view')]");
    private final By createClientButton = By.xpath("//button[contains(., 'Create Client')]");
    private final By loadingSpinner = By.cssSelector(".MuiCircularProgress-root");
    private final By errorAlert = By.cssSelector(".MuiAlert-standardError");
    private final By statsCards = By.cssSelector(".MuiCard-root");

    public ReportsPage(WebDriver driver) {
        super(driver);
    }

    public boolean isReportsPageDisplayed() {
        try {
            waitUtils.waitForElementVisible(pageTitle);
            return isElementDisplayed(pageTitle);
        } catch (Exception e) {
            return false;
        }
    }

    public void waitForPageLoad() {
        waitUtils.waitForElementInvisible(loadingSpinner);
        waitUtils.waitForElementVisible(pageTitle);
        waitUtils.sleep(500);
    }

    public boolean hasNoClients() {
        return isElementDisplayed(noClientsMessage);
    }

    public ClientsPage clickCreateClientButton() {
        click(createClientButton);
        return new ClientsPage(driver);
    }

    public boolean isSelectClientMessageDisplayed() {
        return isElementDisplayed(selectClientMessage);
    }

    public void selectClient(String clientName) {
        click(clientSelect);
        waitUtils.sleep(300);
        List<WebElement> menuItems = getElements(clientMenuItems);
        for (WebElement item : menuItems) {
            if (item.getText().equals(clientName)) {
                item.click();
                waitUtils.sleep(500);
                waitUtils.waitForElementInvisible(loadingSpinner);
                return;
            }
        }
    }

    public void selectClientByIndex(int index) {
        click(clientSelect);
        waitUtils.sleep(300);
        List<WebElement> menuItems = getElements(clientMenuItems);
        if (index < menuItems.size()) {
            menuItems.get(index).click();
            waitUtils.sleep(500);
            waitUtils.waitForElementInvisible(loadingSpinner);
        }
    }

    public void selectFirstClient() {
        selectClientByIndex(1);
    }

    public String getTotalHours() {
        try {
            waitUtils.waitForElementVisible(totalHoursCard);
            return getText(totalHoursCard);
        } catch (Exception e) {
            return "0.00";
        }
    }

    public String getTotalEntries() {
        try {
            waitUtils.waitForElementVisible(totalEntriesCard);
            return getText(totalEntriesCard);
        } catch (Exception e) {
            return "0";
        }
    }

    public String getAverageHours() {
        try {
            waitUtils.waitForElementVisible(averageHoursCard);
            return getText(averageHoursCard);
        } catch (Exception e) {
            return "0.00";
        }
    }

    public boolean areStatsCardsDisplayed() {
        return getElementCount(statsCards) >= 3;
    }

    public void clickCsvExportButton() {
        List<WebElement> buttons = driver.findElements(By.cssSelector("button"));
        for (WebElement btn : buttons) {
            if (btn.getAttribute("innerHTML").contains("DescriptionIcon") ||
                btn.getAttribute("aria-label") != null && btn.getAttribute("aria-label").contains("CSV")) {
                btn.click();
                waitUtils.sleep(1000);
                return;
            }
        }
    }

    public void clickPdfExportButton() {
        List<WebElement> buttons = driver.findElements(By.cssSelector("button"));
        for (WebElement btn : buttons) {
            if (btn.getAttribute("innerHTML").contains("PictureAsPdfIcon") ||
                btn.getAttribute("aria-label") != null && btn.getAttribute("aria-label").contains("PDF")) {
                btn.click();
                waitUtils.sleep(1000);
                return;
            }
        }
    }

    public boolean isCsvExportEnabled() {
        List<WebElement> buttons = driver.findElements(By.cssSelector("button"));
        for (WebElement btn : buttons) {
            if (btn.getAttribute("innerHTML").contains("DescriptionIcon")) {
                return btn.isEnabled() && !btn.getAttribute("class").contains("Mui-disabled");
            }
        }
        return false;
    }

    public boolean isPdfExportEnabled() {
        List<WebElement> buttons = driver.findElements(By.cssSelector("button"));
        for (WebElement btn : buttons) {
            if (btn.getAttribute("innerHTML").contains("PictureAsPdfIcon")) {
                return btn.isEnabled() && !btn.getAttribute("class").contains("Mui-disabled");
            }
        }
        return false;
    }

    public boolean isReportTableDisplayed() {
        return isElementDisplayed(reportTable);
    }

    public int getReportEntryCount() {
        if (isElementDisplayed(noEntriesMessage)) {
            return 0;
        }
        List<WebElement> rows = getElements(reportRows);
        if (rows.size() == 1 && rows.get(0).getText().contains("No work entries")) {
            return 0;
        }
        return rows.size();
    }

    public boolean hasNoWorkEntries() {
        return isElementDisplayed(noEntriesMessage) || getReportEntryCount() == 0;
    }

    public String getErrorAlertText() {
        try {
            return getText(errorAlert);
        } catch (Exception e) {
            return null;
        }
    }

    public boolean isErrorAlertDisplayed() {
        return isElementDisplayed(errorAlert);
    }

    public List<String> getAvailableClients() {
        click(clientSelect);
        waitUtils.sleep(300);
        List<WebElement> menuItems = getElements(clientMenuItems);
        List<String> clients = menuItems.stream()
                .map(WebElement::getText)
                .filter(text -> !text.equals("Choose a client..."))
                .toList();
        driver.findElement(By.tagName("body")).click();
        waitUtils.sleep(200);
        return clients;
    }

    public String getSelectedClient() {
        try {
            return getText(clientSelect);
        } catch (Exception e) {
            return null;
        }
    }
}

File: e2e-tests/page-objects/ReportsPage.js
import BasePage from './BasePage.js';

class ReportsPage extends BasePage {
    get pageHeader() { return $('h4'); }
    get clientSelector() { return $('//label[contains(text(),"Select Client")]/following-sibling::div'); }
    get csvExportButton() { return $('//button[contains(@aria-label,"Export as CSV") or .//svg[contains(@data-testid,"DescriptionIcon")]]'); }
    get pdfExportButton() { return $('//button[contains(@aria-label,"Export as PDF") or .//svg[contains(@data-testid,"PictureAsPdfIcon")]]'); }
    get totalHoursCard() { return $('//p[contains(text(),"Total Hours")]/following-sibling::div'); }
    get totalEntriesCard() { return $('//p[contains(text(),"Total Entries")]/following-sibling::div'); }
    get averageHoursCard() { return $('//p[contains(text(),"Average Hours per Entry")]/following-sibling::div'); }
    get workEntriesTable() { return $('table'); }
    get noClientsMessage() { return $('//p[contains(text(),"You need to create at least one client before generating reports")]'); }
    get selectClientMessage() { return $('//p[contains(text(),"Select a client to view their time report")]'); }
    get noWorkEntriesMessage() { return $('//p[contains(text(),"No work entries found for this client")]'); }
    get createClientButton() { return $('//a[contains(text(),"Create Client")]'); }

    async open() {
        await super.open('/reports');
    }

    async getPageHeader() {
        const header = await this.pageHeader;
        await header.waitForDisplayed();
        return await header.getText();
    }

    async isClientSelectorDisplayed() {
        try {
            const selector = await this.clientSelector;
            return await selector.isDisplayed();
        } catch {
            return false;
        }
    }

    async selectClient(clientName) {
        const selectDiv = await this.clientSelector;
        await selectDiv.click();
        await browser.pause(300);
        const option = await $(`//li[contains(text(),"${clientName}")]`);
        await option.waitForDisplayed();
        await option.click();
        await browser.pause(1000);
    }

    async getTotalHours() {
        const card = await $('//p[contains(text(),"Total Hours")]/ancestor::div[contains(@class,"MuiCardContent")]//div[contains(@class,"MuiTypography-h4")]');
        await card.waitForDisplayed();
        return await card.getText();
    }

    async getTotalEntries() {
        const card = await $('//p[contains(text(),"Total Entries")]/ancestor::div[contains(@class,"MuiCardContent")]//div[contains(@class,"MuiTypography-h4")]');
        await card.waitForDisplayed();
        return await card.getText();
    }

    async getAverageHoursPerEntry() {
        const card = await $('//p[contains(text(),"Average Hours per Entry")]/ancestor::div[contains(@class,"MuiCardContent")]//div[contains(@class,"MuiTypography-h4")]');
        await card.waitForDisplayed();
        return await card.getText();
    }

    async isWorkEntriesTableDisplayed() {
        try {
            const table = await this.workEntriesTable;
            return await table.isDisplayed();
        } catch {
            return false;
        }
    }

    async isWorkEntryRowDisplayed(hours, description) {
        try {
            let selector;
            if (description) {
                selector = `//td//span[contains(text(),"${hours} hours")]/ancestor::tr//p[contains(text(),"${description}")]`;
            } else {
                selector = `//td//span[contains(text(),"${hours} hours")]`;
            }
            const row = await $(selector);
            return await row.isDisplayed();
        } catch {
            return false;
        }
    }

    async clickCsvExportButton() {
        const button = await $('//button[.//svg[@data-testid="DescriptionIcon"]]');
        await button.waitForDisplayed();
        await button.click();
        await browser.pause(1000);
    }

    async clickPdfExportButton() {
        const button = await $('//button[.//svg[@data-testid="PictureAsPdfIcon"]]');
        await button.waitForDisplayed();
        await button.click();
        await browser.pause(1000);
    }

    async isCsvExportButtonDisabled() {
        try {
            const button = await $('//button[.//svg[@data-testid="DescriptionIcon"]]');
            const isDisabled = await button.getAttribute('disabled');
            return isDisabled !== null;
        } catch {
            return true;
        }
    }

    async isPdfExportButtonDisabled() {
        try {
            const button = await $('//button[.//svg[@data-testid="PictureAsPdfIcon"]]');
            const isDisabled = await button.getAttribute('disabled');
            return isDisabled !== null;
        } catch {
            return true;
        }
    }

    async isNoClientsMessageDisplayed() {
        try {
            const message = await this.noClientsMessage;
            return await message.isDisplayed();
        } catch {
            return false;
        }
    }

    async isSelectClientMessageDisplayed() {
        try {
            const message = await this.selectClientMessage;
            return await message.isDisplayed();
        } catch {
            return false;
        }
    }

    async isNoWorkEntriesMessageDisplayed() {
        try {
            const message = await this.noWorkEntriesMessage;
            return await message.isDisplayed();
        } catch {
            return false;
        }
    }

    async isCreateClientButtonDisplayed() {
        try {
            const button = await this.createClientButton;
            return await button.isDisplayed();
        } catch {
            return false;
        }
    }
}

export default new ReportsPage();
